fix: absolute paths for plans

piraces opened the pull request:
Fixes #2298

After looking in detail for the issue, I noticed that CUE is interpreting absolute paths as package names and not as "local packages".

Since the comment that for now dagger only supports local filesystem paths, I changed the funcion `loadPlan` to resolve the given path in arguments to a relative path from current working directory. I also checked if a single file is passed in the arguments to resolve this accordingly.

This relative path resolution allows CUE to load all provided paths as local packages.

Also added support to resolve "~" to current user home. The tilde should be expanded by the shell, Golang does not expand it, so we need to "manually" expand the tilde to the current user home.

piraces commented:
Will have to look for the relative path resolution... Seems like it's failing in some cases 😞

piraces commented:
I have tested it thoughtfully, I find it now a stable solution for interpreting absolute paths regarding the default CUE behaviour with absolute paths. Any feedback will be very much appreciated 😃

TomChv reviewed (commented):
Thanks for your contribution :rocket:

I manually tested the feature according to Solomon's issue:

```
cd /tmp
git clone https://github.com/dagger/dagger

dagger do -p /tmp/dagger --help
dagger project not found. Run `dagger project init`
```

It looks like it do not found the project, I think it come from those lines
```go
	if strings.HasPrefix(absPlanPath, workingDirPath) {
 		relPlanPath, err = filepath.Rel(workingDirPath, absPlanPath)
 	} else {
 		relPlanPath, err = filepath.Rel(absPlanPath, workingDirPath)
 	}
```

There is also an issue with `~` resolution

```shell
file ~/tmp/dagger               
/Users/tomchauveau/tmp/dagger: directory

dagger do -p ~/tmp/dagger --help
stat ../../Documents/DAGGER/dagger: no such file or directory
```

I think there should be some additional tests to verify that everything works

piraces commented:
Thanks, I think I messed up the resolution in the latest commit... I will look further and in detail into it

piraces commented:
I decided to take a simpler approach due to the complexity of making all routes relative and force CUE to interpret them correctly.
This approach makes a Chdir into the absolute path if provided to load the plan and when its loaded, it returns to the original working directory.

I think it's the simpler approach due to current support for only local filesystem paths.
I'm open to suggestions or pointers for other approaches... As I cannot find a better solution 😞

TomChv reviewed (commented):
I still hit the same issue, am I doing something wrong?

```shell
$ pwd
/tmp

$ dagger do -p /tmp/dagger
failed to load plan: dagger project not found. Run `dagger project init`
#...

$ cd dagger 

$ dagger do -p /tmp/dagger
Usage: 
  dagger do  [flags]

#...
Available Actions:
 version FIXME: this can be removed once `go` supports built-in VCS info
```

piraces commented:
@TomChv , what I have been doing in order to test it (Ubuntu 20.04) is the following:

- Building the binary (git repo with this branch in `/tmp/dagger`):
```shell
make dagger
```

- Moving the binary to my executables path:
```shell
sudo cp /tmp/dagger/cmd/dagger/dagger /usr/local/bin/
```
- Testing it:
```shell
piraces:/tmp$ pwd
/tmp
piraces:/tmp$ dagger do -p /tmp/dagger
Usage:
  dagger do  [flags]

Options


Available Actions:
 version FIXME: this can be removed once `go` supports built-in VCS info
 build
 test
 lint
 gitpod


Flags:
      --cache-from stringArray   External cache sources (eg. user/app:cache, type=local,src=path/to/dir)
      --cache-to stringArray     Cache export destinations (eg. user/app:cache, type=local,dest=path/to/dir)
      --dry-run                  Dry run mode
  -h, --help                     help for do
      --no-cache                 Disable caching
      --output string            File path to write the action's output values. Prints to stdout if empty
      --output-format string     Format for output values (plain, json, yaml) (default "plain")
  -p, --plan string              Path to plan (defaults to current directory) (default ".")
      --platform string          Set target build platform (requires experimental)
  -w, --with stringArray

Global Flags:
      --experimental        Enable experimental features
      --log-format string   Log format (auto, plain, tty, json) (default "auto")
  -l, --log-level string    Log level (default "info")
```

```shell
piraces:/tmp$ cd dagger
piraces:/tmp/dagger$ dagger do -p /tmp/dagger
Usage:
  dagger do  [flags]

Options


Available Actions:
 version FIXME: this can be removed once `go` supports built-in VCS info
 build
 test
 lint
 gitpod


Flags:
      --cache-from stringArray   External cache sources (eg. user/app:cache, type=local,src=path/to/dir)
      --cache-to stringArray     Cache export destinations (eg. user/app:cache, type=local,dest=path/to/dir)
      --dry-run                  Dry run mode
  -h, --help                     help for do
      --no-cache                 Disable caching
      --output string            File path to write the action's output values. Prints to stdout if empty
      --output-format string     Format for output values (plain, json, yaml) (default "plain")
  -p, --plan string              Path to plan (defaults to current directory) (default ".")
      --platform string          Set target build platform (requires experimental)
  -w, --with stringArray

Global Flags:
      --experimental        Enable experimental features
      --log-format string   Log format (auto, plain, tty, json) (default "auto")
  -l, --log-level string    Log level (default "info")
```

The output seems correct in my case... I have rebased from main and it's still working. What are the steps you're doing in order to test this branch?

TomChv commented on plan/plan.go:
```diff
@@ -40,11 +40,12 @@ type Config struct {
 	defer span.End()
 
-	log.Ctx(ctx).Debug().Interface("args", cfg.Args).Msg("loading plan")
+	log.Ctx(ctx).Debug().Interface("args", cfg.Args).Msg("loading plan with evaluated args")
+	log.Ctx(ctx).Debug().Interface("path", originalPlan).Msg("loading plan in path")
```
Really good idea IMO but that is only accurate in that PR, you should remove it before merge

piraces replied:
Removed, thanks!

TomChv reviewed (commented):
Ok that works well:
- `dagger do -p /tmp/dagger` in `/tmp` ✅
- `dagger do -p /tmp/dagger` in `/tmp/dagger` ✅

But `dagger do -p ./dagger` in `/tmp` do not work anymore :O 
I do not really understand why because `dagger do -p dagger` (without `./`) works

piraces commented:
> But `dagger do -p ./dagger` in `/tmp` do not work anymore :O I do not really understand why because `dagger do -p dagger` (without `./`) works

It seems that is failing in the main branch too, as I tested.
 
> And that works if I go to `dagger/tests/tasks/exec` and to `dagger do -p ./user.cue test`

It's because it is a single file and CUE hands it correctly. Nevertheless, it appears that the problem with paths and plans are not only with absolute paths, its also with relative routes to directories.

I have changed the approach to `chdir` into relative directories too and make then plan load inside that dir (as I have done with absolute paths).

With this change, that cases are working ok.

TomChv reviewed (commented):
You are right!

To fix this problem, we must make `Dir` of cue loader mutable instead of hard coding it to `""`.
But that add another dimension of complexity in path resolution.
That should be done in another PR.

Otherwise, code looks ok to me.
That would be great to create unit test for the loader.
/cc @aluzzardi

piraces commented:
Yes, it seems to make `Dir` of cue loader mutable instead of hard coding it to `""` its the way to solve this, I agree to solve this in a different PR.

Regarding the unit test for the loader, I totally agree and I can work on that test in this PR.

piraces commented:
I made some unit tests for the logic added to the `loadPlan` function. Essentially, they check that we are changing directory only when its needed. Please let me know if you think more tests or other logic its needed.

TomChv reviewed (approved):
LGTM! Thanks for your amazing follow up!

@aluzzardi @helderco We can merge this one after your approval

shykes commented:
> Yes, it seems to make `Dir` of cue loader mutable instead of hard coding it to `""` its the way to solve this, I agree to solve this in a different PR.

FYI you can find more information on that topic in the discussion thread at https://github.com/dagger/dagger/issues/1620

Thanks for this contribution!

mircubed commented:
@aluzzardi @helderco reminder to review

samalba commented:
This is one looks good to me, @aluzzardi should take a second look given the changes on error reporting.

TomChv commented:
@piraces Do not forget to rebase your PR so we can merge it just after @aluzzardi reviewed it (if it approves it)

piraces commented:
Already rebased @TomChv , sorry for the delay...

The `Test Universe / Universe Tests - Europa (pull_request)` appears to be failing due to a temporal error or downtime in `nuget.org`:

```
Unable to read data from the transport connection: Connection reset by peer.
Connection reset by peer
Failed to download package 'Microsoft.CodeCoverage.16.11.0' from '[https://api.nuget.org/v3-](https://api.nuget.org/v3-flatcontainer/microsoft.codecoverage/16.11.0/microsoft.codecoverage.16.11.0.nupkg)
```

Can you attempt to re-run this check? I think everything should be ok... if not let me know 👍

mircubed commented:
@aluzzardi ready for your review

aluzzardi reviewed (changes requested):
Thanks for the contribution @piraces, much appreciated!

Sorry I left a wall of comments, but I think there's a few design issues that will yield to unexpected problems

piraces commented:
Thanks for the pointers! Working on this...

piraces commented on cmd/dagger/cmd/do.go:
```diff
@@ -230,12 +232,35 @@ var doCmd = &cobra.Command{
-	absPlanPath, err := filepath.Abs(planPath)
+	homePathAlias := fmt.Sprintf("~%c", os.PathSeparator)
+	if strings.HasPrefix(planPath, homePathAlias) {
+		dirname, err := os.UserHomeDir()
```
Correct. I just changed to use `homedir.Expand(planPath)` instead of `os.UserHomeDir`.

piraces commented on cmd/dagger/cmd/do.go:
```diff
@@ -230,12 +232,35 @@ var doCmd = &cobra.Command{
-	_, err = os.Stat(absPlanPath)
+	isDirectory := fileInfo.IsDir()
+
+	if isDirectory && !strings.HasPrefix(planPath, currentPathAlias) && !strings.HasPrefix(planPath, parentPathAlias) {
```
Agree. Incorrect approach from my part... Sorry for this, I was missing the main point of using the CUE `Dir` field as you said in the following conversation. I have changed the approach now and this logic is now not present. Thanks for pointing it out.

piraces commented on cmd/dagger/cmd/do.go:
```diff
@@ -230,12 +232,35 @@ var doCmd = &cobra.Command{
+	isDirectory := fileInfo.IsDir()
+
+	if isDirectory && !strings.HasPrefix(planPath, currentPathAlias) && !strings.HasPrefix(planPath, parentPathAlias) {
+		err = os.Chdir(planPath)
```
Thank you very much for this information. I think this is the main point for the whole bug and the related development. I learned a lot from this. All the logic I was working on can be solved only making use of the CUE `Dir` field.

As you can see in `plan/plan.go` line 56 now I'm doing the following:

```golang
v, err := compiler.Build(ctx, cueModParent, nil, ".")
```

In this way, we are setting the CUE `Dir` field to the `cueModParent` and then loading the plan in a relative way (passing "." for this). This causes CUE to load the plan correctly avoiding to use `os.Chdir` which leads potentially to unexpected behaviour.

piraces replied:
This appears not to be working correctly as I thought. I keep investigating...

piraces commented:
Hi, I made a change in the original approach based on the @aluzzardi review. I have ensured everything its working now. But the PR workflows are being cancelled due to timeout... I have run those test locally without any problem.

Maybe it's something related with my added code? If it is due to this I can't identify what could it be 😢

piraces commented:
I have been working on this. The original approach was incorrect from my part... Sorry for this, I was missing the main point of using the CUE `Dir` field as you said in the following conversation. I have changed the approach now and this logic is now not present. Thanks for pointing it out.

I just changed to use `homedir.Expand(planPath)` instead of `os.UserHomeDir` for tilde resolution in `do.go` and `plan.go`.

Thank you very much for this information. I think this is the main point for the whole bug and the related development. I learned a lot from this. All the logic I was working on can be solved only making use of the CUE Dir field.

As you can see in `plan/plan.go` line 71 now I'm doing the following:
```golang
v, err := compiler.Build(ctx, src, nil, args)
```
In this way, we are setting the CUE `Dir` field to the `cueModParent` if its an absolute path and not the default plan path "." and then loading the plan in a relative way. In order to mantain the original behaviour, checks are performed to mantain the CUE `Dir` path and args (ensuring the original behaviour).

This causes CUE to load the plan correctly avoiding to use `os.Chdir` which leads potentially to unexpected behaviour.

I have tested all the original test cases (with absoulte paths) and they are all working.

Please let me think what do you think about this approach and potential problems if you see them.

gerhard commented:
This is 99% there. The only thing missing is an integration test. I am going to add it next.

In the meantime, do you want to take another look at the code changes @aluzzardi?

gerhard reviewed (approved):
✅ [Integration test passes](https://github.com/dagger/dagger/runs/7434651132?check_suite_focus=true#step:10:1301). This one does something very similar to what was failing in https://github.com/dagger/dagger/issues/2298.

OK to merge from my POV ⛹️‍♂️ WDYT @aluzzardi?

aluzzardi reviewed (approved):
This looks **great**! Thanks for making all the changes!

LGTM

gerhard commented:
Thank you @aluzzardi! Will squash & merge as soon as all checks go green.

gerhard commented:
Thank you @piraces for sticking with us! 💪

We intend to pay closer attention to PRs going forward and have them flowing through quicker. 🏎
